internal/telemetry/collector/segmentio.go

```diff
@@ -0,0 +1,469 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package collector holds different implementation of telemetry data collectors.
+package collector
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/cenkalti/backoff/v4"
+	libsveltosv1beta1 "github.com/projectsveltos/libsveltos/api/v1beta1"
+	"github.com/segmentio/analytics-go/v3"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	kubevirtv1 "kubevirt.io/api/core/v1"
+	clusterapiv1 "sigs.k8s.io/cluster-api/api/v1beta1" // TODO: switch to v1beta2 everywhere
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	kcmv1 "github.com/K0rdent/kcm/api/v1beta1"
+	"github.com/K0rdent/kcm/internal/utils/pointer"
+)
+
+type SegmentIO struct {
+	segmentCl       analytics.Client
+	crCl            client.Client
+	childScheme     *runtime.Scheme
+	expBackoff      backoff.BackOff
+	childFactory    func([]byte, *runtime.Scheme) (client.Client, error) // for test mocks
+	systemNamespace string
+	concurrency     int
+}
+
+type (
+	nodeDataAccumulator struct {
+		nodeInfos     []map[string]string
+		totalCPU      uint64
+		totalMemory   uint64
+		totalGPUNodes uint
+		count         uint
+	}
+
+	podDataAccumulator struct {
+		nodeInfos       *[]map[string]string
+		nodeName2Idx    map[string]int
+		podsWithGPUReqs uint
+	}
+)
+
+func NewSegmentIO(systemNamespace string, segmentClient analytics.Client, crClient client.Client, concurrency int) (*SegmentIO, error) {
+	s := runtime.NewScheme()
+
+	for _, f := range []func(*runtime.Scheme) error{
+		corev1.AddToScheme,
+		metav1.AddMetaToScheme,
+		kubevirtv1.AddToScheme,
+		appsv1.AddToScheme,
+	} {
+		if err := f(s); err != nil {
+			return nil, fmt.Errorf("failed to add to scheme: %w", err)
+		}
+	}
+
+	return &SegmentIO{
+		expBackoff:      backoff.NewExponentialBackOff(backoff.WithInitialInterval(500*time.Millisecond), backoff.WithMaxElapsedTime(10*time.Second)),
+		segmentCl:       segmentClient,
+		crCl:            crClient,
+		systemNamespace: systemNamespace,
+		childScheme:     s,
+		concurrency:     concurrency,
+		childFactory:    defaultClientFactory,
+	}, nil
+}
+
+func (t *SegmentIO) Collect(ctx context.Context) error {
+	l := ctrl.LoggerFrom(ctx)
+
+	mgmt := new(kcmv1.Management)
+	if err := t.retry(ctx, func() error {
+		return t.crCl.Get(ctx, client.ObjectKey{Name: kcmv1.ManagementName}, mgmt)
+	}); err != nil {
+		return fmt.Errorf("failed to get Management: %w", err)
+	}
+
+	mgmtID := string(mgmt.UID)
+
+	templatesList := new(kcmv1.ClusterTemplateList)
```

Comment:
Shouldn’t we retrieve all templates across the entire cluster (in non-system namespace), not just the built-in ones?

Reply:
We should, it was copied and pasted from the original implementation, fixed